internal/entities/properties/service/service.go

```diff
@@ -0,0 +1,267 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package service provides a service to interact with properties of an entity
+package service
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/entities/properties"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+	v1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+const (
+	// propertiesCacheTimeout is the default timeout for the cache of properties
+	propertiesCacheTimeout = 60
+	// bypassCacheTimeout is a special value to bypass the cache timeout
+	// it is not exported from the package and should only be used for testing
+	bypassCacheTimeout = -1
+)
+
+// PropertiesService is the interface for the properties service
+type PropertiesService interface {
+	// RetrieveAllProperties fetches all properties for the given entity
+	RetrieveAllProperties(context.Context, v1.Provider, uuid.UUID, string, minderv1.Entity) (*properties.Properties, error)
```

Comment:
should we retrive the entity by name or upstream ID?

Reply:
should the `Retrieve*` methods _also_ receive properties? it could be a subset of properties that help the provider identify the entity. e.g. only the upstream ID.

Reply:
Interesting! What would the Retrieve method do with the properties? Iterate over them trying to find the identifier to fetch the entity by (we could maybe have cases for name and upstream ID) - but then I guess the method would only save the properties that would be retrieved from upstream and not the ones passed as parameters?

Reply:
uhm... wait a minute.... If this is retrieving all the properties for an entity, wouldn't it suffice to use the actual entity ID? I got a bit confused with the case of the provider fetching properties for an entity Minder doesn't know about yet.

Reply:
Maybe we're talking about different IDs?

The entity might not be cached yet which would generate the ID that is used in the database as the entity's primary key. But the caller must know what entity do they want to refresh - by "an attribute". I read your suggestion as to provide Properties that the code would then iterate over and pick the property that it would fetch the entity from DB by.

Now the code assumes you fetch an entity by name. What I read you suggesting was to pass properties:

```go
fetchBy, _ := properties.NewProperties(map[string]any{
  properties.PropertyName: "stacklok/minder",
}
```

or:
```go
fetchBy, _ := properties.NewProperties(map[string]any{
  properties.PropertyUpstreamID: "this-can-even-be-a-uuid",
}
```

and then the Retrieve method would do something like:
```go
for name, value := range fetchBy.Iterate() {
 if name == properties.PropertyName {
   db.FetchEntityByName()
 } else if name == properties.PropertyUpstreamID. {
   db.FetchEntityByUpstreamID()
 }
}
```

or did you mean something else?